### PR TITLE
Fix file uploading

### LINKED ISSUE #316: Upload Error 404
When I try upload file, i got error

**error sending message: document upload failed: upload responsed with %!d(float64=404)**

but I can send text message with no problem

### PR DESCRIPTION
**Pull request by tulir**
Fixes #316 

WhatsApp changed something again. I have no idea if I did this right, but it seems to work.

### CODE REVIEW

**Comment by luwice** [comment]
Thanks, It works

**Comment by x00b** [comment]
Is this branch going to be merged?

**Comment by gvn182** [comment]
@tulir Thanks, that works! Can you tell us how did you get this solutions?

**Comment by tulir** [comment]
Basically just looked at what web.whatsapp.com did using firefox devtools:

* query `mediaConn` is used, action `encr_upload` isn't, so the former is probably the replacement for the latter. It's sent unencrypted in the websocket, so you can easily see the request and response in the websocket message viewer.
* the upload request itself only contained random bytes, so I guessed it's the raw encrypted data instead of multipart form data like before.
* the URL was just first hostname returned by mediaConn and there was `/mms/image` after that. Not sure if the other hostnames that mediaConn return should be used in some cases.
* there was no content-type header or `f=j` query param like before, so I removed those.
* `auth` query param was the auth field returned by `mediaConn`.
* `token` param looked like url-safe base64 and the old code had a base64 of the encrypted file sha256, so I tried passing that as the token and it worked.

**Comment by Ased2235** [comment]
is it me or thumbnails isn't working?

**Comment by sthales** [comment]
> is it me or thumbnails isn't working?

Its working fine here.

**Comment by Mateus-pilo** [comment]
Has this update been updated on the master yet?

**Comment by beshoo** [comment]
Is it working with thumbnails! and when it will be available on master

**Comment by beshoo** [comment]
@tulir
1- Is it working with thumbnails! 
2- What about audio and other documents type? 
since:

> the URL was just first hostname returned by mediaConn and there was /mms/image after that. Not sure if the other hostnames that mediaConn return should be used in some cases.

`Path:     fmt.Sprintf("/mms/image/%s", token)`  why don't you use what filetype function which it is return the correct media path?

Whatsapp is using the following mms paths so please use the filetype function to determinate the correct path.
```
            audio: "/mms/audio",
            document: "/mms/document",
            gif: "/mms/gif",
            image: "/mms/image",
            ppic: "/pps/photo",
            product: "/mms/image",
            ptt: "/mms/ptt",
            sticker: "/mms/sticker",
            video: "/mms/video"
```

**Comment by tulir** [comment]
Other file types are working fine even with the wrong path (at least audio and documents), it's not like WhatsApp can see what type of data is inside the file. But yeah, those paths would make sense so I'll change it to use those.

**Comment by beshoo** [comment]
@Rhymen hi.
I would like to ask you when this update will going to be merged

File: media.go [hunk 1]
```diff
@@ -128,67 +169,30 @@ func (wac *Conn) Upload(reader io.Reader, appInfo MediaType) (url string, mediaK
...
-		filetype = "document"
-	case MediaVideo:
-		filetype = "video"
-	}
-
-	uploadReq := []interface{}{"action", "encr_upload", filetype, base64.StdEncoding.EncodeToString(fileEncSha256)}
-	ch, err := wac.writeJson(uploadReq)
-	if err != nil {
-		return "", nil, nil, nil, 0, err
+	hostname, auth, _, err := wac.queryMediaConn()
```
Comment (AlmogBaku) [inline media.go:172]: don't we want to cache the token by the TTL?

Comment (beshoo) [reply media.go:172]: @tulir what do you think?

Comment (tulir) [reply media.go:172]: Works fine without caching too, but it can be added later if someone else wants to do it

**Comment by p4kl0nc4t** [comment]
When will this get merged?

**Comment by Rhymen** [comment]
Thanks for the amazing work!

**Comment by beshoo** [comment]
Dear all, i still get an error 404 while sending media.
Media sent successfully, but this strange, why i got this error....

**Comment by Piyushhbhutoria** [comment]
**Please update the latest release with these changes. 🙂** 

It took me 3 hrs to debug to find out that the version go get was fetching was the latest release and not from the master. 
The current `latest-release` was done on `1st Jan 2020`. But this `error fix was merged` on `8th Apr 2020`. Please release a new update as this is breaking my app.

**Comment by thomcatdotrocks** [comment]
@Piyushhbhutoria Can you just clone the repo?

As for file uploading, I've noticed I'm still unable to send GIFs, not sure if that's in the pipeline or if I should submit a new feature request.

**Pull request merged**